monai/handlers/decollate_batch.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING
+
+from monai.config import IgniteInfo
+from monai.data import decollate_batch, rep_scalar_to_batch
+from monai.engines.utils import IterationEvents
+from monai.utils import min_version, optional_import
+
+Events, _ = optional_import("ignite.engine", IgniteInfo.OPT_IMPORT_VERSION, min_version, "Events")
+if TYPE_CHECKING:
+    from ignite.engine import Engine
+else:
+    Engine, _ = optional_import("ignite.engine", IgniteInfo.OPT_IMPORT_VERSION, min_version, "Engine")
+
+
+class DecollateBatch:
+    """
+    Ignite handler to execute the `decollate batch` logic for `engine.state.batch` and `engine.state.output`.
+    Typical usage is to set `decollate=False` in the engine and execute some postprocessing logic first
+    then decollate the batch, otherwise, engine will decollate batch before the postprocessing.
+
+    Args:
+        event: expected EVENT to attach the handler, should be "MODEL_COMPLETED" or "ITERATION_COMPLETED".
+            default to "MODEL_COMPLETED".
+
+    """
+
+    def __init__(self, event: str = "MODEL_COMPLETED"):
+        event = event.upper()
+        if event not in ("MODEL_COMPLETED", "ITERATION_COMPLETED"):
+            raise ValueError("event should be `MODEL_COMPLETED` or `ITERATION_COMPLETED`.")
+        self.event = event
+
+    def attach(self, engine: Engine) -> None:
+        """
+        Args:
+            engine: Ignite Engine, it can be a trainer, validator or evaluator.
+        """
+        if self.event == "MODEL_COMPLETED":
+            engine.add_event_handler(IterationEvents.MODEL_COMPLETED, self)
+        else:
+            engine.add_event_handler(Events.ITERATION_COMPLETED, self)
```

Comment:
should the `PostProcessing` handler have the same logic of choosing the event to attach?

or perhaps this one  is not needed, the user can have `monai.handlers.PostProcessing(transform=Decollated)`

Reply:
Hi @wyli ,

I also added the logic to choose event in PostProcessing handler in this PR.
But I think that's a good idea to use `Decollated` transform directly, any other use-case to define this handler?

Thanks.

Reply:
I think maybe we can't put `Decollated` in a transform chain of `monai.handlers.PostProcessing` because `engine_apply_transform` will temporarily combine 2 dictionaries and output 2 dictionaries, it can't support the `Decollated` transform which output a list of dictionaries, unless we customize a logic in `Postprocessing` handler for `Decollated` transform.
Let's keep to use a handler for `decollate` first?

Thanks.

Reply:
sure, we can refactor this later